src/Mvc/test/Mvc.IntegrationTests/ValidationWithRecordIntegrationTests.cs

```diff
@@ -2298,6 +2298,119 @@ public async Task Validation_InifnitelyRecursiveModel_ValidationOnTopLevelParame
+            var parameterBinder = ModelBindingTestHelper.GetParameterBinder(modelMetadataProvider);
+            var expected = $"Record type '{modelType}' has validation metadata defined on property 'Property1' that will be ignored. " +
+                "Consider specifying these on the constructor parameter 'Property1' instead.";
+
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "parameter",
+                ParameterType = modelType,
+            };
+
+            var testContext = ModelBindingTestHelper.GetTestContext(request =>
+            {
+                request.QueryString = new QueryString("?Property1=8");
+            });
+
+            var modelState = testContext.ModelState;
+
+            // Act & Assert
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(() =>
+                parameterBinder.BindModelAsync(parameter, testContext, modelMetadataProvider, modelMetadata));
+
+            Assert.Equal(expected, ex.Message);
+        }
+
+        private record RecordTypeWithValidatorsOnProperties(string Property1)
+        {
+            [Required]
+            public string Property1 { get; init; }
+        }
+
+        [Fact]
+        public async Task Validation_ValidatorsDefinedOnRecordTypePropertiesAndParameters()
+        {
+            // Arrange
+            var modelType = typeof(RecordTypeWithValidatorsOnPropertiesAndParameters);
+            var modelMetadataProvider = TestModelMetadataProvider.CreateDefaultProvider();
+            var modelMetadata = modelMetadataProvider.GetMetadataForType(modelType);
+            var parameterBinder = ModelBindingTestHelper.GetParameterBinder(modelMetadataProvider);
+            var expected = $"Record type '{modelType}' has validation metadata defined on property 'Property1' that will be ignored. " +
+                "Consider specifying these on the constructor parameter 'Property1' instead.";
+
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "parameter",
+                ParameterType = modelType,
+            };
+
+            var testContext = ModelBindingTestHelper.GetTestContext(request =>
+            {
+                request.QueryString = new QueryString("?Property1=8");
+            });
+
+            var modelState = testContext.ModelState;
+
+            // Act & Assert
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(() =>
+                parameterBinder.BindModelAsync(parameter, testContext, modelMetadataProvider, modelMetadata));
+
+            Assert.Equal(expected, ex.Message);
+        }
+
+        private record RecordTypeWithValidatorsOnPropertiesAndParameters([Required] string Property1)
+        {
+            [Required]
+            public string Property1 { get; init; }
+        }
+
+        [Fact]
+        public async Task Validation_ValidatorsDefinedOnMixOfRecordTypePropertiesAndParameters()
```

Comment:
Do we already have tests covering mix of constructor parameter metadata and validation metadata on properties not mentioned in the primary constructor❔ Those should succeed.

Reply:
I don't think there was one before, I've added one now.